### PR TITLE
support globs in artifact path

### PR DESCRIPTION
**Pull request by michaelglass**
~~# 🛑 not ready for merge until captain will accept `original_path`~~

allows globs in the `path:` field of `upload-captain-artifact`

### CODE REVIEW

File: src/run.ts [hunk 1]
```diff
@@ -33,7 +34,20 @@ export default async function run(): Promise<void> {
...
+          if (globsWithFiles.length !== 0) {
+            return [
+              withFiles.concat(
+                globsWithFiles.map(file => ({
+                  ...artifact,
+                  path: file
+                }))
+              ),
+              withoutFiles
+            ]
```
Comment (kylekthompson) [inline src/run.ts:47]: One thing to note- we'll need to re-work some of the test results displays and flaky tests to merge artifacts within a single job which have the same name.

In other words, if I have 10 tests in file 1 and 15 in file 2, we'll probably just want to show "25 xUnit results" or whatever (that'll additionally be important for the flaky test detection code, which compares test results from artifacts across two jobs)

Comment (kylekthompson) [reply src/run.ts:47]: We'll probably want to do the server-side work before we publish this to v1

Comment (dan-manges) [reply src/run.ts:47]: Could we provide both an aggregated view and an individual view? I could see some scenarios where people would want to aggregate the test results, and others where they'd want to keep them separated. I was going to suggest adding a configuration option like `aggregate: true|false`, but if the UI is built to easily view aggregated or disaggregated, then we don't need to configure anything on upload.

Comment (kylekthompson) [reply src/run.ts:47]: For displaying in the UI, that would be possible (though we probably want to store the original file name, so it’s clear what’s from what. Right now we only have the name)

For flaky tests, I’d worry we’ll have misses if parallel distribution puts results in different files so I’d probably still prefer if those were all rolled up

**Comment by michaelglass** [comment]
@kylekthompson what's the uuid on an artifact used for? Do we want a single uuid per artifact? per glob?

**Comment by michaelglass** [comment]
can we safely group by artifact name?

**Comment by michaelglass** [comment]
I guess rather than a UUID I'd rather pass
- file path for artifacts
- requested path (e.g. the glob) for grouping id.

Thoughts?

File: package-lock.json [hunk 2]
```diff
@@ -2598,7 +2599,6 @@
       "version": "2.1.5",
       "resolved": "https://registry.npmjs.org/@nodelib/fs.scandir/-/fs.scandir-2.1.5.tgz",
       "integrity": "sha512-vq24Bq3ym5HEQm2NKCr3yXDwjc7vTsEThRDnkp2DK9p1uqLR+DHurm/NOTo0KG7HYHU7eppKZj3MyqYuMBf62g==",
-      "dev": true,
```
Comment (michaelglass) [inline package-lock.json:2601]: I have no idea why this happened 👎

File: package-lock.json [hunk 3]
```diff
@@ -11,6 +11,7 @@
       "dependencies": {
         "@actions/core": "1.9.1",
         "@actions/github": "5.1.0",
+        "fast-glob": "3.2.12",
```
Comment (michaelglass) [inline package-lock.json:14]: I don't have a strong opinion about the library. The main alternative might be https://www.npmjs.com/package/glob

**Comment by dan-manges** [comment]
Re: the UUID discussion, my understanding is that we're using this string to match the artifacts and make a subsequent API call that the artifact has been uploaded. I think it should be safe to use the file path, assuming that the identifiers are only used within the span of a single execution of `upload-captain-artifact`. However, I could see somebody having multiple uploads within the same job that are actually different file paths:

```
job:
  name: whatever
  steps:
    - run: cd subdir1 && bundle exec rspec --write-to-path rspec.json
    - uses: rwx-research/upload-captain-artifact
      with:
        -path: rspec.json
    - run: cd subdir2 && bundle exec rspec --write-to-path rspec.json
    - uses: rwx-research/upload-captain-artifact
      with:
        -path: rspec.json
```

I'm pretty sure that usage will still be fine, but the slight risk that we have ambiguous identifiers and mess something up makes me feel inclined to continue using random UUIDs. If It's completely behind the scenes, we don't really gain anything by swapping the random identifier with a specific one. That said, I could also see passing up the glob pattern and/or specific file path since I imagine that metadata around the test results could be helpful in display or for diagnostics.

**Comment by michaelglass** [comment]
helpful. I assumed a fixed working directory.

**Pull request merged**